pkg/tnf/pkg/pacemaker/healthcheck.go

```diff
@@ -0,0 +1,691 @@
+			window = eventDeduplicationWindowFencing
+		}
+
+		// Remove if expired
+		if now.Sub(timestamp) > window {
+			delete(c.recordedEvents, key)
+		}
+	}
+}
+
+// shouldRecordEvent checks if an event should be recorded based on deduplication logic.
+// This should be called after cleanupExpiredEvents() has been called.
+func (c *HealthCheck) shouldRecordEvent(eventKey string, deduplicationWindow time.Duration) bool {
+	c.recordedEventsMu.Lock()
+	defer c.recordedEventsMu.Unlock()
+
+	now := time.Now()
+
+	// Check if this event was recently recorded
+	if lastRecorded, exists := c.recordedEvents[eventKey]; exists {
+		if now.Sub(lastRecorded) < deduplicationWindow {
+			return false
+		}
+	}
+
+	// Record this event
+	c.recordedEvents[eventKey] = now
+	return true
+}
+
+// recordHealthCheckEvents records events for pacemaker warnings and fencing history
+func (c *HealthCheck) recordHealthCheckEvents(status *HealthStatus) {
+	// Clean up expired events before processing new ones
+	c.cleanupExpiredEvents()
+
+	// Record events for warnings with appropriate deduplication window
+	for _, warning := range status.Warnings {
+		eventKey := fmt.Sprintf(eventKeyPrefixWarning+"%s", warning)
+		// Use longer deduplication window for fencing events
+		deduplicationWindow := eventDeduplicationWindowDefault
+		if strings.Contains(warning, warningPrefixFencingEvent) {
+			deduplicationWindow = eventDeduplicationWindowFencing
+		}
+		if c.shouldRecordEvent(eventKey, deduplicationWindow) {
+			c.recordWarningEvent(warning)
+		}
+	}
+
+	// Record events for errors with default deduplication window
+	for _, err := range status.Errors {
+		eventKey := fmt.Sprintf(eventKeyPrefixError+"%s", err)
+		if c.shouldRecordEvent(eventKey, eventDeduplicationWindowDefault) {
+			c.eventRecorder.Warningf(eventReasonError, msgPacemakerError, err)
+		}
+	}
+
+	// Record a normal event for healthy status only when transitioning from unhealthy/unknown
+	if status.OverallStatus == statusHealthy {
+		var wasUnhealthy bool
+		func() {
+			c.previousStatusMu.Lock()
+			defer c.previousStatusMu.Unlock()
+			wasUnhealthy = c.previousStatus != statusHealthy
+			c.previousStatus = statusHealthy
+		}()
+
+		if wasUnhealthy {
+			c.eventRecorder.Eventf(eventReasonHealthy, msgPacemakerHealthy)
+		}
+	} else {
+		// Update previous status for non-healthy states
+		func() {
+			c.previousStatusMu.Lock()
+			defer c.previousStatusMu.Unlock()
+			c.previousStatus = status.OverallStatus
+		}()
+	}
+}
+
+// recordWarningEvent records appropriate warning events based on warning type
+func (c *HealthCheck) recordWarningEvent(warning string) {
+	switch {
+	case strings.Contains(warning, warningPrefixFailedAction):
+		c.eventRecorder.Warningf(eventReasonFailedAction, msgDetectedFailedAction, warning)
+	case strings.Contains(warning, warningPrefixFencingEvent):
+		c.eventRecorder.Warningf(eventReasonFencingEvent, msgDetectedFencing, warning)
+	default:
```

Comment:
Nitpick: as the warning already contains the prefix ("Recent failed resource action:"), I think the event message will have it twice